File: LeetCode/1727. Largest Submatrix With Rearrangements/solution.py
class Solution:
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        n, m = len(matrix), len(matrix[0])
        dp = [[0]*m for _ in range(n+1)]
        res = 0
        for i in range(n):
            arr = []
            for j in range(m):
                if matrix[i][j]:
                    dp[i][j] = dp[i-1][j] + matrix[i][j]
                arr.append(dp[i][j])
            arr = sorted(arr, key=lambda x: -x)
            # print(arr)
            h = float('inf')
            for j in range(m):
                h = min(h,arr[j])
                res = max(res,(j+1)*h)
        # print(dp)
        return res 

File: LeetCode/1727. Largest Submatrix With Rearrangements/solution.cpp
class Solution {
public:
    int largestSubmatrix(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size(), res = 0;
        vector<int> cc(m,0);
        for (int i = 0; i < n; i++){
            vector<int> height;
            for (int j = 0; j < m; j++){
                cc[j] = matrix[i][j] == 0 ? 0 : cc[j]+1;
                height.push_back(cc[j]);
            }
            sort(height.begin(), height.end(), greater<>());
            int h = n;
            for (int j = 0; j < m && height[j] > 0; j++){
                h = min(h, height[j]);
                res = max(res, h*(j+1));
            }
        }
        return res;
    }
};